src/graphics/shader.rs

```diff
@@ -53,77 +180,59 @@ use wgpu::util::DeviceExt;
 /// ```
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct Shader {
-    pub(crate) fragment: ArcShaderModule,
-    pub(crate) fs_entry: String,
+    pub(crate) vs_module: Option<ArcShaderModule>,
+    pub(crate) fs_module: Option<ArcShaderModule>,
 }
 
-impl Shader {
-    /// Creates a shader from a WGSL string.
-    pub fn from_wgsl(gfx: &impl Has<GraphicsContext>, wgsl: &str, fs_entry: &str) -> Self {
-        let gfx = gfx.retrieve();
-        let module = ArcShaderModule::new(gfx.wgpu.device.create_shader_module(
-            wgpu::ShaderModuleDescriptor {
-                label: None,
-                source: wgpu::ShaderSource::Wgsl(wgsl.into()),
-            },
-        ));
+pub use crevice::std140::AsStd140;
 
-        Shader {
-            fragment: module,
-            fs_entry: fs_entry.into(),
-        }
-    }
+/// A builder for [ShaderParams]
+#[derive(Debug)]
+pub struct ShaderParamsBuilder<'a, Uniforms: AsStd140> {
+    uniforms: &'a Uniforms,
+    images: &'a [(&'a Image, Sampler)],
+    images_vs_visible: bool,
 }
 
-pub use crevice::std140::AsStd140;
+impl<'a, Uniforms: AsStd140> ShaderParamsBuilder<'a, Uniforms> {
+    /// Creates a new builder for [ShaderParams].
+    ///
+    /// # Arguments
+    ///
+    /// * `uniforms` - Initial uniforms.
+    pub fn new(uniforms: &'a Uniforms) -> Self {
+        ShaderParamsBuilder {
+            uniforms,
+            images: &[],
+            images_vs_visible: false,
+        }
+    }
 
-/// Parameters that can be passed to a custom shader, including uniforms, images, and samplers.
-///
-/// These parameters are bound to group 4. With WGSL, for example,
-/// ```rust,ignore
-/// ggez::graphics::ShaderParams::new(ctx, &my_uniforms, &[&image1, &image2], &[sampler1])
-/// ```
-/// Corresponds to...
-/// ```ignore
-/// @group(4) @binding(0)
-/// var<uniform> my_uniforms: MyUniforms;
-/// @group(4) @binding(1)
-/// var image1: texture_2d<f32>;
-/// @group(4) @binding(2)
-/// var image2: texture_2d<f32>;
-/// @group(4) @binding(3)
-/// var sampler1: sampler;
-/// ```
-#[derive(Debug, PartialEq, Eq)]
-pub struct ShaderParams<Uniforms: AsStd140> {
-    pub(crate) uniforms: ArcBuffer,
-    pub(crate) layout: ArcBindGroupLayout,
-    pub(crate) bind_group: ArcBindGroup,
-    _marker: PhantomData<Uniforms>,
-}
+    /// Provides images to the shaders.
+    ///
+    /// # Arguments
+    ///
+    /// * `vs_visible` - If the images should also be visible to the vertex shader, rather
+    ///    than just the fragment shader.
+    pub fn images(self, images: &'a [(&'a Image, Sampler)], vs_visible: bool) -> Self {
+        ShaderParamsBuilder {
+            uniforms: self.uniforms,
+            images,
+            images_vs_visible: vs_visible,
+        }
+    }
```

Comment:
Ah, there's been a misunderstanding what sampler means in relation to images. You should *not* have 1 sampler to 1 image, so they shouldn't be bundled at all. It would be expected you reuse the same sampler in the shader for sampling N images. The most common use case is to have 1 sampler and N (>1) images. Of course you can have more than 1 sampler if you wish to sample differently, e.g., in the same shader you want to sample both linearly and nearest (which is quite a rare thing to do). Regardless, they should be taken separately and stored separately as you may have a different number of each -- however, if this at all helps with deciding API ergonomics, when you want to make the images vertex-visible, you almost certainly also want to make the samplers vertex-visible.

Reply:
Ah thanks! Okay I separated the lists again (I think that should be sufficient).

Reply:
Hmm. On one hand it's more straightforward to nudge/remind the user to hand over samplers (at least one) as well when handing over images. On the other hand we already have a builder and we could easily split this up into seperate functions for handing over images and samplers each (the concept of users having to hand over slices of them is probably worth keeping, as it makes it easier for users to see and control the order of samplers/images which they need to know while writing the shader).

But in the end I too think that this is sufficient.

Reply:
could we pass the canvas's current sampler by default perhaps?

Reply:
Nah, let's better not hide some implicit magic here. Explicit is clearer and not too ugly imo.

Reply:
I'll push a separate branch if that's alright.